sympy/core/expr.py

```diff
@@ -329,98 +329,80 @@ def __complex__(self):
         re, im = result.as_real_imag()
         return complex(float(re), float(im))
 
-    def __ge__(self, other):
-        from sympy import GreaterThan
+    def _cmp(self, other, op, cls):
+        assert op in ("<", ">", "<=", ">=")
         try:
             other = _sympify(other)
         except SympifyError:
-            raise TypeError("Invalid comparison %s >= %s" % (self, other))
+            raise TypeError("Invalid comparison %s %s %s" % (self, op, other))
         for me in (self, other):
             if me.is_extended_real is False:
                 raise TypeError("Invalid comparison of non-real %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
+
         n2 = _n2(self, other)
         if n2 is not None:
-            return _sympify(n2 >= 0)
+            # use float comparison for infinity.
+            # otherwise get stuck in infinite recursion
+            if n2 in (S.Infinity, S.NegativeInfinity):
+                n2 = float(n2)
```

Comment:
This condition wasn't here before. Why is it needed now?

Reply:
Before, infinity defined its own comparison operators that didn't call _n2. Much of that code duplicated what is done here, so I merged them together. Basically, the problem is that n2 is expected to return either a Float or None for comparable objects. Since infinity is comparable, it doesn't return None, but `Float('inf')` is the same as Infinity which is not actually a Float. So `n2 op 0` calls `_cmp` recursively forever

Reply:
This wouldn't have been needed in older versions where `Float('inf')` was distinct from `S.Infinity`, but I assume that was changed for a reason